src/main/scala-3/com/raquo/airstream/split/SplittableTypeMacros.scala

```diff
@@ -0,0 +1,185 @@
+    handlerMapExpr: Expr[Map[Int, Function[Any, O]]],
+    casePfExpr: Expr[PartialFunction[A, B]],
+    handleFnExpr: Expr[Function[(B, Signal[B]), O1]]
+  )(
+    using quotes: Quotes
+  ): Expr[MatchSplitObservable[Self, I, O1]] = {
+    val caseExprList = exprOfListToListOfExpr(caseListExpr)
+
+    val nextCaseExprList = casePfExpr.asExprOf[PartialFunction[Any, Any]] :: caseExprList
+
+    val nextCaseListExpr = listOfExprToExprOfList(nextCaseExprList)
+
+    '{ MatchSplitObservable.build[Self, I, O1]($observableExpr, $nextCaseListExpr, ($handlerMapExpr + ($handlerMapExpr.size -> $handleFnExpr.asInstanceOf[Function[Any, O1]]))) }
+  }
+
+  private def exprOfListToListOfExpr(
+    pfListExpr: Expr[List[PartialFunction[Any, Any]]]
+  )(
+    using quotes: Quotes
+  ): List[Expr[PartialFunction[Any, Any]]] = {
+    import quotes.reflect.*
+    
+    pfListExpr match {
+      case '{ $headExpr :: (${tailExpr}: List[PartialFunction[Any, Any]]) } =>
+        headExpr :: exprOfListToListOfExpr(tailExpr)
+      case '{ Nil } => Nil
+      case _ => report.errorAndAbort("Macro expansion failed, please use `handleCase` instead of modify MatchSplitObservable explicitly")
+    }
+    
+  }
+
+  private def listOfExprToExprOfList(
+    pfExprList: List[Expr[PartialFunction[Any, Any]]]
+  )(
+    using quotes: Quotes
+  ): Expr[List[PartialFunction[Any, Any]]] = {
+    import quotes.reflect.*
+    
+    pfExprList match
+      case head :: tail => '{ $head :: ${listOfExprToExprOfList(tail)} }
+      case Nil => '{ Nil }
+  }
+
+  private inline def toSplittableOneObservable[Self[+_] <: Observable[_], O](
+    parentObservable: BaseObservable[Self, (Int, Any)],
+    handlerMap: Map[Int, Function[Any, O]]
+  ): Self[O] = {
+    parentObservable.matchStreamOrSignal(
+      ifStream = _.splitOne(_._1) { case (idx, (_, b), dataSignal) =>
+        val bSignal = dataSignal.map(_._2)
+        handlerMap.apply(idx).apply(b -> bSignal)
+      },
+      ifSignal = _.splitOne(_._1) { case (idx, (_, b), dataSignal) =>
+        val bSignal = dataSignal.map(_._2)
+        handlerMap.apply(idx).apply(b -> bSignal)
+      }
+    ).asInstanceOf[Self[O]] // #TODO[Integrity] Same as FlatMap/AsyncStatusObservable, how to type this properly?
+  }
+
+  private def observableImpl[Self[+_] <: Observable[_] : Type, I: Type, O: Type](
+    matchSplitObservableExpr: Expr[MatchSplitObservable[Self, I, O]]
+  )(
+    using quotes: Quotes
+  ): Expr[Self[O]] = {
+    import quotes.reflect.*
+
+    matchSplitObservableExpr match {
+      case '{ MatchSplitObservable.build[Self, I, O]($_, Nil, $_)} =>
+        report.errorAndAbort("Macro expansion failed, need at least one handleCase")
+      case '{ MatchSplitObservable.build[Self, I, O]($observableExpr, $caseListExpr, $handlerMapExpr)} => 
+        '{toSplittableOneObservable(
+          $observableExpr
+            .map(i => ${ innerObservableImpl('i, caseListExpr) })
+            .asInstanceOf[BaseObservable[Self, (Int, Any)]],
```

Comment:
Is this `asInstanceOf` actually needed? Which part of the type is unproven? Is it due to Airstream types, or something inherent in macros?

I would expect the output here to be `Self[(Int, Any)]` prior to `asInstanceOf`. I'm not sure how `(Int, Any)` relates to `O`, but I don't see how `asInstanceOf` helps with that either, since it does not mention`O`. It seems that you're using `asInstanceOf` to get `BaseObservable`, but all you need to return is Self, no? I could be missing something...

Reply:
`$observableExpr.map(i => ${ innerObservableImpl('i, caseListExpr) })` is indeed a `Self[(Int, Any)]`, but I'm not returning this, I'm actually feeding this into `toSplittableOneObservable` function call, and in that function I need `matchStreamOrSignal`, so `Self` is not enough.